tools/generate-examples.rb

```diff
@@ -0,0 +1,409 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require "fileutils"
+require "pathname"
+
+GraalVMOld = Struct.new(:url, :dir) do
+  def name
+    dir
+  end
+
+  def tarball
+    @tarball ||= File.basename(url)
+  end
+
+  def install
+    system("curl", "-OL", url) unless File.exist?(tarball)
+    FileUtils.mkdir_p(name)
+
+    Dir.chdir(name) do
+      FileUtils.mkdir_p("#{__dir__}/../examples/#{name}")
+
+      bin = "#{dir}/Contents/Home/bin"
+
+      log(self, "Installing Truffle languages.")
+
+      unless Dir.exist?(dir)
+        system("tar", "-zxf", "../#{tarball}")
+        system("#{bin}/gu", "install", "nodejs")
+        system("#{bin}/gu", "install", "ruby")
+      end
+    end
+  end
+
+  def java_context(&block)
+    Dir.chdir(name) do
+      block.call("#{dir}/Contents/Home/bin")
+    end
+  end
+
+  alias_method :js_context, :java_context
+  alias_method :ruby_context, :java_context
+end
+
+GraalVM = Struct.new(:java_version, :truffle_version, :community_edition, :dir) do
+  def name
+    dir
+  end
+
+  def install
+    FileUtils.mkdir_p(name)
+
+    Dir.chdir(name) do
+      FileUtils.mkdir_p("#{__dir__}/../examples/#{name}")
+
+      # Install the JDK.
+      log(self, "Installing JDK.")
+      system("curl", "-OL", java_download_url) unless File.exist?(File.basename(java_download_url))
+
+      unless Dir.exist?("jdk")
+        FileUtils.mkdir("jdk")
+        system("tar", "-zxf", File.basename(java_download_url), "-C", "jdk", "--strip-components=1")
+      end
+
+      # Install GraalJS.
+      log(self, "Installing GraalJS.")
+      system("curl", "-OL", js_download_url) unless File.exist?(File.basename(js_download_url))
+
+      unless Dir.exist?("js")
+        FileUtils.mkdir("js")
+        system("tar", "-zxf", File.basename(js_download_url), "-C", "js", "--strip-components=1")
+      end
+
+      # Install TruffleRuby.
+      log(self, "Installing TruffleRuby.")
+      system("curl", "-OL", ruby_download_url) unless File.exist?(File.basename(ruby_download_url))
+
+      unless Dir.exist?("ruby")
+        FileUtils.mkdir("ruby")
+        system("tar", "-zxf", File.basename(ruby_download_url), "-C", "ruby", "--strip-components=1")
+      end
```

Comment:
This would make a tidy function that could be reused for the next two :)
```suggestion
    def install_project(name, dir, url)
      basename = File.basename(url)
      log(self, "Installing #{name}.")
      system("curl", "-OL", url) unless File.exist?(basename)

      unless Dir.exist?(dir)
        FileUtils.mkdir(dir)
        system("tar", "-zxf", basename, "-C", dir, "--strip-components=1")
      end
    end
    install_project("JDK", "jdk", java_download_url)
    install_project("GraalJS", "js", js_download_url)
    install_project("TruffleRuby", "ruby", ruby_download_url)
```

Reply:
That's a good suggestion. This was originally written as a slightly smarter shell script. I've obviously updated a lot of it, but I was also trying to avoid a full rewrite.